event_model/__init__.py

```diff
@@ -632,6 +457,250 @@ def __call__(self, name, doc, validate=False):
+        # keyed on EventDescriptor UID.
+        self._subfactory_cbs_by_descriptor = defaultdict(list)
+
+        # Callbacks that want documents related to a given EventDescriptor,
+        # keyed on the RunStart UID referenced by that EventDescriptor.
+        self._subfactory_cbs_by_start = defaultdict(list)
+
+        # Map RunStart UID to the list EventDescriptor. This is used to
+        # facilitate efficient cleanup of the caches above.
+        self._descriptors = defaultdict(list)
+
+        # Map EventDescriptor UID to RunStart UID. This is used for looking up
+        # Fillers.
+        self._descriptor_to_start = {}
+
+        # Map Resource UID to RunStart UID.
+        self._resources = {}
+
+        # Old-style Resources that do not have a RunStart UID
+        self._unlabeled_resources = deque(maxlen=10000)
+
+        # Map Runstart UID to instances of self.filler_class.
+        self._fillers = {}
+
+    def __repr__(self):
+        return (f"RunRouter([\n" +
+                f"\n".join(f"    {factory}" for factory in self.factories) +
+                f"])")
+
+    def start(self, doc):
+        uid = doc['uid']
+        filler = self.filler_class(self.handler_registry,
+                                   root_map=self.root_map,
+                                   inplace=False)
+        self._fillers[uid] = filler
+        # No need to pass the document to filler
+        # because Fillers do nothing with 'start'.
+        for factory in self.factories:
+            callbacks, subfactories = factory('start', doc)
+            self._factory_cbs_by_start[uid].extend(callbacks)
+            self._subfactories[uid].extend(subfactories)
+
+    def descriptor(self, doc):
+        uid = doc['uid']
+        start_uid = doc['run_start']
+        self._fillers[start_uid].descriptor(doc)
+        # Apply all factory cbs for this run to this descriptor, and run them.
+        factory_cbs = self._factory_cbs_by_start[start_uid]
+        self._factory_cbs_by_descriptor[uid].extend(factory_cbs)
+        for callback in factory_cbs:
+            callback('descriptor', doc)
+        # Let all the subfactories add any relavant callbacks.
+        for subfactory in self._subfactories[start_uid]:
+            callbacks = subfactory('descriptor', doc)
+            self._subfactory_cbs_by_start[start_uid].extend(callbacks)
+            self._subfactory_cbs_by_descriptor[uid].extend(callbacks)
+        # Keep track of the RunStart UID -> [EventDescriptor UIDs] mapping for
+        # purposes of cleanup in stop().
+        self._descriptors[start_uid].append(uid)
+        # Keep track of the EventDescriptor UID -> RunStartUID for filling
+        # purposes.
+        self._descriptor_to_start[uid] = start_uid
+
+    def event_page(self, doc):
+        descriptor_uid = doc['descriptor']
+        start_uid = self._descriptor_to_start[descriptor_uid]
+        try:
+            doc = self._fillers[start_uid].event_page(doc)
```

Comment:
This line can also raise UnresolvableForeignKeyError. Should we handle this exception also?

Reply:
Good question. I think cannot ignore that exception, given when it is raised. Suppose that I make `RunRouter([...], handler_registry={'SOME_SPEC': ...})`. If I then give that RunRouter an Event that references some `datum_id` that it doesn't recognize, it raise `UnresolvableForeignKeyError`. At that moment, I don't yet know whether the data relates to `'SOME_SPEC'` or not because I haven't gotten to the step where we have the Resource yet. Since the user is expecting all data from `'SOME_SPEC'` to be filled, I can't ignore the `UnresolvableForeignKeyError` and let the data pass through unfilled.

So, in effect, this PR is making the RunRouter stricter: it's adding validation. But the constraint that it is enforcing is one that we should be enforcing.